### PR TITLE
Per-function requirements

### LINKED ISSUE #29: Support per-function requirements
It would be nice if different functions could have different requirements

example:

```
├── serverless.yml
├── function1
│      ├── requirements.txt
│      └── index.py
└── function2
       ├── requirements.txt
       └── index.py
```

### LINKED ISSUE #119: Is it possible to specify a different requirement.txt file per function?
Is it possible to specify a different requirement.txt file per function in the serverless.yml function?
Currently all my functions are getting inflated because of requirements needed by another specific function.

### PR DESCRIPTION
**Pull request by cgrimal**
First version of the per-function requirements support.

Still work in progress, but I'd be grateful if I can get some feedback already.

**Summary:**
- requirements are downloaded in separated folder (name of the function)
- symlinks are made in the folder of each function (instead of the root of the project)
- all files are excluded by default, and only the folder of the function is added

The include/exclude per-function were ignoring the requirements files, I implemented a first version, but it is not supporting file globbing yet.

This requires quite some tests, especially because I am not sure it is not breaking the support of some options.

**TODO:**
- ~~**Implement the unlinking of files!**~~
- ~~**Implement artifact creation to avoid subfolder path issue**~~
- Implement file globbing for include/exclude at function level (maybe this is another feature)
- Make sure the other options are still working
  - dockerizePip, dockerSsh, dockerImage and dockerFile are OK
  - ~~**zip is broken**~~ zip is OK

**Related issues:**
- Closes #29 
- Closes #119

### CODE REVIEW

**Comment by dschep** [comment]
I'm out skiing for the weekend so I'll check this out more thoroughly when I get back.

I'd assumed that there would be per function packaging hooks that this would leverage, is that not the case?

**Comment by xerfuz** [comment]
Merge

**Comment by dschep** [comment]
I would prefer if this leveraged function level `package` options instead of adding the `module` key and added a function level `requirementsFileName`? IE:
```yaml
package:
  individually: true
  exclude:
    - '**/*'

functions:
  func1:
    handler: function1/index.handler
    requirementsFileName: function1/requirements.txt
    package:
      include:
        - 'function1/*'
  func2:
    handler: function2/index.handler
    requirementsFileName: function2/requirements.txt
    package:
      include:
        - 'function2/*'
```

That way the only custom `serverless-python-requirements` option is the function level `requirementsFileName`. All the other package individually stuff would then be standard sls options.

**Comment by cgrimal** [comment]
Hi @dschep and thanks for your feedback!

I made quite a lot of changes since you made it though, but I did not really take it into account yet.

That said, I am not sure I really agree with your proposal as you have to repeat `function1` quite a lot.
The only `serverless-python-requirements` option I am adding is the `module` one, so it is not very different from adding `requirementsFileName`.

On top of that, having the path to the module folder is very important IMO as it tells you where to link the requirements, what to zip when you create the artifact (it wasn't the case in my first version). So if I used your example, I would have to extract the module folder from the `requirementsFileName` and it wouldn't be as explicit.

FYI, I used it to deploy my current project with 3 different functions, 2 of them sharing the same code with different environment variables, and it works like a charm!

Here is an example:
```yml
package:
  individually: true

functions:
  func1:
    handler: index.handler
    module: function1
  func2:
    handler: index.handler
    module: function2
```

**Comment by dschep** [comment]
Cool! I'm fine with this way if it doesn't cause unexpected behavior when using function level `package` directives.

**Comment by dschep** [comment]
I'll review the code and test it over the weekend. I'd love it if you could add a test case for this. The tests are in `test.bats` (using [bats](https://github.com/sstephenson/bats))

**Comment by cgrimal** [comment]
Just added a test in the bats file. I used a dedicated test project in `test-indiv`.

Let me know once you have tested it!

File: lib/zip.js [hunk 1]
```diff
@@ -33,22 +60,56 @@ function addVendorHelper() {
...
+        .forEach((f) => {
+          if (!doneModules.includes(f.module)) {
+            this.serverless.cli.log(`Zipping required Python packages for ${f.module}...`);
+            f.package.include.push(`${f.module}/.requirements.zip`);
+            zipper
+              .sync
+              .zip(`.serverless/${f.module}/requirements`)
+              .compress()
+              .save(`${f.module}/.requirements.zip`);
+            doneModules.push(f.module);
```
Comment (ryansb) [inline lib/zip.js]: Nice - I like the caching on this to avoid re-packing deps for modules that have multiple entry points

File: lib/zip.js [hunk 2]
```diff
@@ -33,22 +60,56 @@ function addVendorHelper() {
  */
 function removeVendorHelper() {
   if (this.options.zip && this.options.cleanupZipHelper) {
-    this.serverless.cli.log('Removing Python requirements helper...');
-    return fse.removeAsync(path.join(this.servicePath, 'unzip_requirements.py'));
+    if (this.serverless.service.package.individually) {
```
Comment (ryansb) [inline lib/zip.js]: This makes ziphelper setting stay global, which I think is ok, but a note for future devs: if you have a need for some functions to self-unzip, you will need to make this per-function.

**Comment by ryansb** [comment]
Looks good to me!

**Comment by dschep** [comment]
Thanks for the review @ryansb :+1:

@cgrimal  I'll probably merge this tonight and make a release.

**Comment by dschep** [comment]
It seems zip(and i think the existing `package.individually` implementation, i think this is because you're using `module` to modify  packaging directives, per [my comment](https://github.com/UnitedIncome/serverless-python-requirements/pull/131#issuecomment-359966620)) isn't working: https://circleci.com/gh/UnitedIncome/serverless-python-requirements/238

I had to add a package.json to get your test to work, but it seems fine once that was added.

I also update circle, so any updates you push to this PR should build automatically too.

**Comment by dschep** [comment]
It seems some of the tests fail for unrelated reasons (`hug` now requires py3k it seems), so I'll fix those issues.

**Comment by cgrimal** [comment]
Hey @dschep and @ryansb, thanks for your help and feedback!
I fixed some of the tests this morning (an error in the zip command) but the pipenv related tests are still ko. I'll have a look at them later but I'm less sure I'll manage to understand what's wrong as I never used pipenv previously.

**Comment by dschep** [comment]
Ah, yeah, I'm not surprised it doesn't work with pipenv, the way that works in the plugin is that if a `Pipfile` is detected, it runs `pipenv lock -r` to generate a `requirements.txt` file that it stores in `.serverless`. To fix it you'd probably have to update the `pipenv.js` module use your new `module` flag to load the `Pipfile` from that module. I'm also ok with making this feature not compatible with pipenv if it's documented and a meaningful error is raised.

**Comment by cgrimal** [comment]
Indeed you are right, I managed to fix the test cases using `pipenv` but now, still 2 tests related to the `package.individually` option are KO...

What would you want to be the expected behaviour if `package.individually` is true, but no `module` flags are provided because there is only one `requirements.txt` file in the root folder?

1. Raise an error: IMO it only makes sense to use this option if you want the artifacts to have different contents.
2. Assume the `module` folder is just the root folder and have individual artifacts with the same content (as before).

**Comment by dschep** [comment]
I think the 2nd option. on that note, does this change work with function level package directives? ie:
```yaml
#blablabla
functions:
  func1:
    handler: index.handler
  func2:
    handler: index.handler
    module: separately-pkgd
    package:
      individually: true
```

Tho the more scenarios i look at the more i like my initial suggestion (abeit repetitive) of using package directives heavily and specifying a `requirements.txt` file per function

**Comment by cgrimal** [comment]
Hi @dschep, I implemented the 2nd solution as per your recommendation so now, all tests are passing, without any modification. 😃 

To answer your question, the trigger to install requirements in separate modules is triggered by the `individually` option only at the top-level. That said, in your example, you would still end up with separate artefacts, but the one of the func1 would contain all requirements. To be honest, I don't really find the test case though... If you want, you can write a couple of test cases, and I'll make sure they pass.

**Comment by dschep** [comment]
Yeah i didn't have a test case for that one. I'll whip when I get the chance. Looking forward to merging this. I think I might release it as `4.0.0-beta` for a while when I merge this since it's a pretty major change.

**Comment by cgrimal** [comment]
Hi @dschep!
Do you have an estimated time of merge or are you waiting on anything on my side?
I am in the middle of writing the documentation for my project that is heavily dependant on this feature, so I just want to know how to handle it.
Thanks!

**Comment by dschep** [comment]
Hey, sorry, been busy with other things. Merged and releasing as `4.0.0-beta`.

**Comment by FelschR** [comment]
I've tried to use this but I get the following error:

```sh
Serverless: Installing requirements of requirements.txt in .serverless...

  Error --------------------------------------------------

  ENOENT: no such file or directory, stat 'C:\[my-project-root]\requirements.txt'

     For debugging logs, run again after setting the "SLS_DEBUG=*" environment variable.
```

For some reason it still tries to read a requirements.txt in my project root folder and not in the function folder.

**Comment by FelschR** [comment]
I've looked into the origin of the error and found that it was caused because it processes one of my non-python lambda functions in `installAllRequirements()`.

I'll create a pull request soon with my changes.

**Comment by amitm02** [comment]
Hi, is this implemented yet? is there an example yml file for the syntax?

**Pull request merged**